# Capturando e tratando exceções

* No exemplo abaixo, a instrução **new FileWriter("OutFile.txt")** invoca o construtor da classe **FileWriter**, que inicializa um _stream_ de saída em um arquivo. Se o arquivo não puder ser aberto, uma exceção do tipo **IOException** é lançada.
* Como essa é uma exceção que pode ser antecipada pelo desenvolvedor e da qual o programa pode se recuperar, trata-se de uma **checked exception**.
* Nesse caso, é necessário cumprir o requisito de captura ou especificação, caso contrário a aplicação não vai compilar. 

```
import java.io.*;
import java.util.List;
import java.util.ArrayList;

public class ListOfNumbers {

    private List<Integer> list;
    private static final int SIZE = 10;

    public ListOfNumbers () {
        list = new ArrayList<Integer>(SIZE);
        for (int i = 0; i < SIZE; i++) {
            list.add(new Integer(i));
        }
    }

    public void writeList() {
	// The FileWriter constructor throws IOException, which must be caught.
        PrintWriter out = new PrintWriter(new FileWriter("OutFile.txt"));

        for (int i = 0; i < SIZE; i++) {
            // The get(int) method throws IndexOutOfBoundsException, which must be caught.
            out.println("Value at: " + i + " = " + list.get(i));
        }
        out.close();
    }
}
```

* Outro ponto que merece atenção no código acima é o laço que percorre a lista de inteiros **list**. 
* Por um descuido do desenvolvedor, o laço alcança um índice maior do que o tamanho da lista, causando uma exceção **unchecked** do tipo **IndexOutOfBoundsException**. 
* Essa exceção, contudo, só acontece durante a execução do programa, e -portanto- não exige o requisito de captura e especificação.
* Dessa forma, a aplicação compila, mas durante sua execução a falta de tratamento para a exceção provocará o termino da execução.

